src/v/cluster/scheduling/leader_balancer_greedy.h

```diff
@@ -94,126 +55,22 @@ class greedy_balanced_shards final : private leader_balancer_strategy {
-                  curr_error, lowest_reassign.from, lowest_reassign.to);
-                if (new_error + error_jitter < curr_error) {
-                    return lowest_reassign;
-                }
-            }
-        }
+    find_movement(const absl::flat_hash_set<raft::group_id>& skip) final {
+        _mi->update_muted_groups(skip);
+        return _even_shard_load_c.recommended_reassignment();
+    }
 
-        return std::nullopt;
+    void apply_movement(const reassignment& r) final {
+        _even_shard_load_c.update_index(r);
     }
 
     std::vector<shard_load> stats() const final {
-        std::vector<shard_load> ret;
-        ret.reserve(_load.size());
-        std::transform(
-          _load.cbegin(),
-          _load.cend(),
-          std::back_inserter(ret),
-          [](const auto& e) {
-              // oddly, absl::btree::size returns a signed type
-              return shard_load{
-                e->first, static_cast<size_t>(e->second.size())};
-          });
-        return ret;
+        return _even_shard_load_c.stats();
     }
 
 private:
-    size_t num_groups() const {
-        return std::accumulate(
-          _cores.cbegin(),
-          _cores.cend(),
-          size_t{0},
-          [this](auto acc, const auto& e) {
-              /*
-               * we aren't going to attempt to move leadership between muted
-               * nodes, so we only count groups with leaders on non-muted nodes
-               * when calculating error / target load.
-               */
-              if (_muted_nodes.contains(e.first.node_id)) {
-                  return acc;
-              }
-              return acc + e.second.size();
-          });
-    }
-
-    size_t num_cores() const {
-        return std::count_if(
-          _cores.cbegin(), _cores.cend(), [this](const auto& core) {
-              return !_muted_nodes.contains(core.first.node_id);
-          });
-    }
-
-    /*
-     * build the load index, which is a vector of iterators to each element in
-     * the core index, where the iterators in the load index are sorted by the
-     * number of groups having their leader on a given core.
-     */
-    void rebuild_load_index() {
-        _load.clear();
-        _load.reserve(_cores.size());
-        _load_map.reserve(_cores.size());
-        for (auto it = _cores.cbegin(); it != _cores.cend(); ++it) {
-            _load.push_back(it);
-            _load_map.emplace(it->first, it->second.size());
-        }
-        std::sort(_load.begin(), _load.end(), [](const auto& a, const auto& b) {
-            return a->second.size() < b->second.size();
-        });
-    }
-
-    index_type _cores;
-    absl::flat_hash_set<model::node_id> _muted_nodes;
-    size_t _num_cores;
-    size_t _num_groups;
-    std::vector<index_type::const_iterator> _load;
-    absl::flat_hash_map<model::broker_shard, size_t> _load_map;
+    std::unique_ptr<leader_balancer_types::muted_index> _mi;
```

Comment:
nit:  Curious about the choice of unique_ptr for this, it is only used within this class IIUC. (feel free to skip changing if its too much work, I'm just curious).